services/venice-controller/src/main/java/com/linkedin/venice/controller/StoreGraveyardCleanupService.java

```diff
@@ -0,0 +1,107 @@
+  private final Time time = new SystemTime();
+  private boolean stop = false;
+
+  public StoreGraveyardCleanupService(
+      VeniceParentHelixAdmin admin,
+      VeniceControllerMultiClusterConfig multiClusterConfig) {
+    this.admin = admin;
+    this.multiClusterConfig = multiClusterConfig;
+    this.sleepIntervalBetweenListFetchMinutes =
+        multiClusterConfig.getGraveyardCleanupSleepIntervalBetweenListFetchMinutes();
+    this.cleanupThread = new Thread(new StoreGraveyardCleanupTask(), "StoreGraveyardCleanupTask");
+  }
+
+  @Override
+  public boolean startInner() throws Exception {
+    cleanupThread.start();
+    return true;
+  }
+
+  @Override
+  public void stopInner() throws Exception {
+    stop = true;
+    cleanupThread.interrupt();
+  }
+
+  private class StoreGraveyardCleanupTask implements Runnable {
+    @Override
+    public void run() {
+      LOGGER.info("Started running {}", getClass().getSimpleName());
+      while (!stop) {
+        try {
+          time.sleep((long) sleepIntervalBetweenListFetchMinutes * Time.MS_PER_MINUTE);
+          // loop all clusters
+          for (String clusterName: multiClusterConfig.getClusters()) {
+            VeniceControllerConfig clusterConfig = multiClusterConfig.getControllerConfig(clusterName);
+            boolean cleanupEnabled = clusterConfig.isStoreGraveyardCleanupEnabled();
+            int delayInMinutes = clusterConfig.getStoreGraveyardCleanupDelayMinutes();
+            if (!cleanupEnabled || !admin.isLeaderControllerFor(clusterName)) {
+              // Only clean up when config is enabled and this is the leader controller for current cluster
+              continue;
+            }
+            // List all stores from graveyard for current cluster
+            List<String> storeNames = admin.getStoreGraveyard().listStoreNamesFromGraveyard(clusterName);
+            for (String storeName: storeNames) {
+              boolean didCleanup = false;
+              try {
+                Stat stat = new Stat();
+                admin.getStoreGraveyard().getStoreFromGraveyard(clusterName, storeName, stat);
+                if ((time.getMilliseconds() - stat.getMtime()) / Time.MS_PER_MINUTE > delayInMinutes) {
+                  admin.removeStoreFromGraveyard(clusterName, storeName);
+                  didCleanup = true;
+                }
+              } catch (ResourceStillExistsException resourceStillExistsException) {
+                // Ignore ResourceStillExistsException which indicates store graveyard is not ready for removal.
+              } catch (Exception exception) {
+                LOGGER.error(
+                    "Encountered exception while handling store graveyard cleanup for store: {} in cluster: {}",
+                    storeName,
+                    clusterName,
+                    exception);
+              }
+              if (didCleanup) {
+                time.sleep(Time.MS_PER_MINUTE);
```

Comment:
Why do we need a minute's sleep after each store?

Reply:
The sleep is to avoid a burst of store graveyard removals.

Reply:
Yeah, I don't mind it. After entering steady state, it will not be an issue